File: src/main/java/io/github/freshsupasulley/dbot/polls/NaturalizePoll.java
package io.github.freshsupasulley.dbot.polls;

import io.github.freshsupasulley.dbot.Main;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;

public class NaturalizePoll extends Poll<NaturalizePoll> {
	
	private final long member;
	
	public NaturalizePoll(TextChannel channel, Member member)
	{
		// 3 day cooldown
		super(0.75f, 5, 259200000L, "Naturalize " + member.getEffectiveName() + "? They will be able to participate in democracy. If failed, they cannot be renaturalized.", channel);
		
		this.member = member.getIdLong();
	}
	
	@Override
	public boolean isDuplicate(NaturalizePoll poll)
	{
		return member == poll.member;
	}
	
	@Override
	protected void pollPassed(JDA jda)
	{
		Guild guild = jda.getGuildById(Main.SERVER_ID);
		guild.retrieveMemberById(member).queue(result ->
		{
			Main.server.naturalize(result);
		}, e ->
		{
			Main.log.error("Failed to find member to naturalize while running naturalization poll", e);
		});
	}
	
	@Override
	protected void pollFailed(JDA jda)
	{
		Main.server.addToCitizenBlacklist(member);
	}
}

File: src/polls/naturalize-poll.ts
import { globalState } from '..';
import { APIUser } from 'discord-api-types/v10';
import BasePoll from './poll';
import { InteractionResponseType } from 'discord-interactions';
import { Exclude, Expose } from 'class-transformer';

export default class NaturalizePoll extends BasePoll {
	userID: string;
	// Doesn't seem to do anything but whatever
	@Exclude()
	username: string;

	constructor(userID: string, username?: APIUser) {
		super('naturalize', 0.75, 5, 259200000);
		this.userID = userID;
		this.username = username?.username ?? "if you're reading this I fucked up";
	}

	firePoll() {
		this.question = `Naturalize ${this.username}? They will be able to participate in democracy. If failed, they cannot be renaturalized.`;
		return super.firePoll();
	}

	isDuplicate(sample: NaturalizePoll): boolean {
		return this.userID == sample.userID;
	}

	async pollPassed() {
		return globalState.naturalize(this.userID);
	}

	async pollFailed() {
		return globalState.addToCitizenBlacklist(this.userID);
	}
}
